runtime/src/installed_scheduler_pool.rs

```diff
@@ -0,0 +1,529 @@
+    type TransactionWithIndex<'tx>: WithTransactionAndIndex;
+}
+
+#[derive(Debug)]
+pub struct DefaultScheduleExecutionArg;
+
+impl ScheduleExecutionArg for DefaultScheduleExecutionArg {
+    type TransactionWithIndex<'tx> = &'tx (&'tx SanitizedTransaction, usize);
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum WaitReason {
+    // most normal termination waiting mode; couldn't be done implicitly inside Bank::freeze() -> () to return
+    // the result and timing in some way to higher-layer subsystems;
+    TerminatedToFreeze,
+    TerminatedFromBankDrop,
+    TerminatedInternallyByScheduler,
+    // scheduler will be restarted without being returned to pool in order to reuse it immediately.
+    ReinitializedForRecentBlockhash,
+}
+
+pub type InstalledSchedulerBox = Box<dyn InstalledScheduler<DefaultScheduleExecutionArg>>;
+// somewhat arbitrary new type just to pacify Bank's frozen_abi...
+#[derive(Debug, Default)]
+pub(crate) struct InstalledSchedulerBoxInBank(Option<InstalledSchedulerBox>);
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+use solana_frozen_abi::abi_example::AbiExample;
+
+#[cfg(RUSTC_WITH_SPECIALIZATION)]
+impl AbiExample for InstalledSchedulerBoxInBank {
+    fn example() -> Self {
+        Self(None)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct SchedulingContext {
+    mode: SchedulingMode,
+    // Intentionally not using Weak<Bank> for performance reasons
+    bank: Arc<Bank>,
+}
+
+impl WithSchedulingMode for SchedulingContext {
+    fn mode(&self) -> SchedulingMode {
+        self.mode
+    }
+}
+
+impl SchedulingContext {
+    pub fn new(mode: SchedulingMode, bank: Arc<Bank>) -> Self {
+        Self { mode, bank }
+    }
+
+    pub fn bank(&self) -> &Arc<Bank> {
+        &self.bank
+    }
+
+    pub fn slot(&self) -> Slot {
+        self.bank().slot()
+    }
+
+    pub fn log_prefix(context: Option<&Self>, scheduler_id: SchedulerId) -> String {
+        const BITS_PER_HEX_DIGIT: usize = 4;
+
+        format!(
+            "sch_{:0width$x}{}: ",
+            scheduler_id,
+            context
+                .map(|c| format!("(slot:{}/mode:{:?})", c.slot(), c.mode()))
+                .unwrap_or_else(|| "(?)".into()),
+            width = SchedulerId::BITS as usize / BITS_PER_HEX_DIGIT,
+        )
+    }
+}
+
+// tiny wrapper to ensure to call schedule_termination() via ::drop() inside
+// BankForks::set_root()'s pruning.
+pub(crate) struct BankWithScheduler(Arc<Bank>);
+
+impl BankWithScheduler {
+    pub(crate) fn new(bank: Arc<Bank>) -> Self {
+        Self(bank)
+    }
+
+    pub(crate) fn bank_cloned(&self) -> Arc<Bank> {
+        self.0.clone()
+    }
+
+    pub(crate) fn bank(&self) -> &Arc<Bank> {
+        &self.0
+    }
+
+    pub(crate) fn into_bank(self) -> Arc<Bank> {
+        let bank = self.bank_cloned();
+        drop(self);
+        bank
+    }
+}
+
+impl Drop for BankWithScheduler {
+    fn drop(&mut self) {
+        self.0.schedule_termination();
```

Comment:
drop-chain for `Bank` and this related type seem complicated. I'm not sure I understand the relationships, and why we need these to be complicated.

Maybe related to my previous comment about where the scheduler lives.

Reply:
yeah, as you spotted this particular line, this is pivotal for scheduler termination. but lacks a clarifying comment. (I'll add one)